components/tidb_query/src/rpn_expr/impl_arithmetic.rs

```diff
@@ -498,21 +486,17 @@ impl ArithmeticOpWithCtx for DecimalDivide {
     type T = Decimal;
 
     fn calc(ctx: &mut EvalContext, lhs: &Decimal, rhs: &Decimal) -> Result<Option<Decimal>> {
-        use crate::codec::mysql::Res;
-
-        Ok(match lhs / rhs {
-            Some(value) => match value {
-                Res::Ok(value) => Some(value),
-                Res::Truncated(_) => ctx.handle_truncate(true).map(|_| None)?,
-                Res::Overflow(_) => ctx
-                    .handle_overflow_err(Error::overflow(
-                        "DECIMAL",
-                        &format!("({} / {})", lhs, rhs),
-                    ))
-                    .map(|_| None)?,
-            },
-            None => ctx.handle_division_by_zero().map(|_| None)?,
-        })
+        if let Some(value) = lhs / rhs {
+            value
+                .into_result_with_overflow_err(
+                    ctx,
+                    Error::overflow("DECIMAL", format!("({} / {})", lhs, rhs)),
+                )
+                .map(Some)
+        } else {
+            ctx.handle_division_by_zero().map(|_| None)
+        }
+        .map_err(|err| err.into())
```

Comment:
Why not `?` directly?

Reply:
If we use `?`, we need to wrap the result with `Ok`.